### PR TITLE
feat(nav): add dropdown menu items

### PR DESCRIPTION
**Pull request by natemoo-re**
- Adds **Help** menu to the nav sidebar added in #77665
- Introduces a generic system for nav sidebar items which should open an overlay menu (will support broadcasts, onboarding)

### CODE REVIEW

File: static/app/components/nav/utils.tsx [hunk 1]
```diff
@@ -102,12 +104,14 @@ export function makeLinkPropsFromTo(to: string): {
   state: object;
   to: LocationDescriptor;
 } {
-  const [pathname, search] = to.split('?');
+  const [rawPathname, search] = to.split('?');
+  const [pathname, hash] = rawPathname.split('#');
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:108]: Why are we doing this and not using [`URL`](https://developer.mozilla.org/en-US/docs/Web/API/URL)

File: static/app/components/nav/sidebar.tsx [hunk 2]
```diff
@@ -89,37 +93,85 @@ const SidebarItemList = styled('ul')`
...
+        {children}
+      </StyledButton>
+      {isOpen && (
+        <OverlayMenu {...overlayProps}>
+          <Suspense fallback={<LoadingIndicator />}>
+            {item.overlay ? <item.overlay /> : null}
+          </Suspense>
+        </OverlayMenu>
+      )}
+    </Fragment>
```
Comment (evanpurkhiser) [inline static/app/components/nav/sidebar.tsx:171]: Why can't this whole thing just use our `DropdownMenu` component?

https://github.com/getsentry/sentry/blob/master/static/app/components/dropdownMenu/index.tsx

File: static/app/components/nav/config.tsx [hunk 3]
```diff
@@ -209,6 +212,11 @@ export function createNavConfig({organization}: {organization: Organization}): N
       {label: t('Alerts'), to: `/${prefix}/alerts/rules/`, icon: <IconSiren />},
     ],
     footer: [
+      {
+        label: t('Help'),
+        icon: <IconQuestion />,
+        overlay: lazy(() => import('./overlay/help')),
```
Comment (evanpurkhiser) [inline static/app/components/nav/config.tsx:218]: Why have lazy loading for this?

File: static/app/components/nav/overlay.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import {type ComponentProps, forwardRef} from 'react';
```
Comment (evanpurkhiser) [inline static/app/components/nav/overlay.tsx:1]: minor nit: You pretty much never need to import react types, since they're all available on the `React` global. So this can just be `React.ComponentProps`

File: static/app/components/nav/utils.tsx [hunk 5]
```diff
@@ -24,7 +24,9 @@ export interface NavItemLayout<Item extends NavSidebarItem | NavSubmenuItem> {
 /** SidebarItem is a top-level NavItem which is always displayed in the app sidebar */
 export interface NavSidebarItem extends NavItem {
   /** The icon to render in the sidebar */
-  icon: JSX.Element;
+  icon: ReactElement;
+  /** if te items to display when this SidebarItem is active */
+  overlay?: LazyExoticComponent<any>;
   /** Optionally, the submenu items to display when this SidebarItem is active */
   submenu?: NavSubmenuItem[] | NavItemLayout<NavSubmenuItem>;
   /**
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:32]: small nit: we don't have a lint for this, but our general docblock style is pretty much always multi line

```
/**
 * Comment here
 */
```
and never 

```
/** Comment here */
```

File: static/app/components/nav/utils.tsx (hunk 1 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx]: Originally because these were path strings and not fully formed URL strings, but great point! I switched this over in 25a82c04624

File: static/app/components/nav/sidebar.tsx (hunk 2 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: Migrated this in [e69afcc](https://github.com/getsentry/sentry/pull/79720/commits/e69afcc68bdf675e2f1f6044e054ad9573d6cf6b)

File: static/app/components/nav/config.tsx (hunk 3 above)
Comment (natemoo-re) [reply static/app/components/nav/config.tsx]: Switched the configuration to a POJO in [e69afcc](https://github.com/getsentry/sentry/pull/79720/commits/e69afcc68bdf675e2f1f6044e054ad9573d6cf6b)

File: static/app/components/nav/utils.tsx (hunk 5 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx]: Reformatted these in f2279cbcd00

File: static/app/components/nav/overlay.tsx (hunk 4 above)
Comment (saponifi3d) [reply static/app/components/nav/overlay.tsx:1]: Hmm - I would actually say it's better to deconstruct it from the import.

By doing that it creates a clear dependency for those properties. I am also extremely weary of the global space, it can often be polluted and cause more harm than good to rely on those.

Comment (natemoo-re) [reply static/app/components/nav/overlay.tsx:1]: I always forget about the `React` global! Updated in 6a06fcea1e5

Comment (JonasBa) [inline static/app/components/nav/overlay.tsx:1]: Just as a word of caution, but whenever we use ComponentProps<typeof Component>, we end up creating a unnecessary type alias with typeof Component, which over time will slow down the type checking and increase memory usage of the language server. Since we dont really have a convention, this is obv fine, but we should really just be importing the component props directly

File: static/app/components/nav/overlay.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import styled from '@emotion/styled';
+
+import InteractionStateLayer from 'sentry/components/interactionStateLayer';
+import Link from 'sentry/components/links/link';
+import {Overlay} from 'sentry/components/overlay';
+import {space} from 'sentry/styles/space';
+
+export const OverlayMenu = forwardRef<
```
Comment (JonasBa) [inline static/app/components/nav/overlay.tsx:9]: I'm scarred from all the generic naming we have in the codebase, but how about?
```suggestion
export const NavOverlayMenu = forwardRef<
```

File: static/app/components/nav/overlay.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  Omit<ComponentProps<typeof StyledOverlay>, `on${string}`>
+>(({children, ...props}, ref) => {
+  return (
+    <StyledOverlay {...props} ref={ref}>
+      {children}
+    </StyledOverlay>
+  );
+});
+
+export const OverlayMenuLink = forwardRef<HTMLAnchorElement, ComponentProps<typeof Link>>(
```
Comment (JonasBa) [inline static/app/components/nav/overlay.tsx:20]: Are these going to be used in a followup pr?

File: static/app/components/nav/utils.tsx [hunk 8]
```diff
@@ -97,17 +119,23 @@ export function isSubmenuItemActive(
...
+ ** Creates a `LocationDescriptor` from a URL string that may contain search params
+ */
 export function makeLinkPropsFromTo(to: string): {
   state: object;
   to: LocationDescriptor;
 } {
-  const [pathname, search] = to.split('?');
+  const {pathname, search, hash} = new URL(
+    to,
+    !to.startsWith('http') ? 'https://sentry.io/' : undefined
```
Comment (JonasBa) [inline static/app/components/nav/utils.tsx:131]: Should we make this a constant somewhere? It looks like something that we'd want to be able to easily change

File: static/app/components/nav/sidebar.tsx [hunk 9]
```diff
@@ -1,9 +1,11 @@
-import {Fragment} from 'react';
+import {Fragment, type PropsWithChildren} from 'react';
```
Comment (JonasBa) [inline static/app/components/nav/sidebar.tsx:1]: A word of caution, but PropsWithChildren doesn't mean that the children are actually required, which I think is what you actually want here?
```
type PropsWithChildren<P = unknown> = P & { children?: ReactNode | undefined };
```

For example in this case, you want the children to always be present right?
```
  return (
    <StyledLink
      {...linkProps}
      className={isActive || isSubmenuActive ? 'active' : undefined}
      aria-current={isActive ? 'page' : undefined}
    >
      <InteractionStateLayer hasSelectedBackground={isActive || isSubmenuActive} />
      {children}
    </StyledLink>
```

File: static/app/components/nav/overlay.tsx (hunk 4 above)
Comment (evanpurkhiser) [reply static/app/components/nav/overlay.tsx:1]: > I am also extremely weary of the global space, it can often be polluted and cause more harm than good to rely on those.

I certainly agree that global space isn't usually great. But I think for the case of react types, especially because they're namespaced under `React`. This is convenient and has very few downsides.

File: static/app/components/nav/config.tsx [hunk 10]
```diff
@@ -209,6 +212,34 @@ export function createNavConfig({organization}: {organization: Organization}): N
       {label: t('Alerts'), to: `/${prefix}/alerts/rules/`, icon: <IconSiren />},
     ],
     footer: [
+      {
+        label: t('Help'),
+        icon: <IconQuestion />,
+        menu: [
```
Comment (evanpurkhiser) [inline static/app/components/nav/config.tsx:218]: Nice!

Comment (evanpurkhiser) [inline static/app/components/nav/config.tsx:218]: This feature might get tricky for some of the menus that sometimes use getsnetry hooks, but we can cross that bridge later 😅

File: static/app/components/nav/overlay.tsx
Comment (evanpurkhiser) [inline static/app/components/nav/overlay.tsx:1]: Is this file used at all anymore?

File: static/app/components/nav/sidebar.tsx [hunk 11]
```diff
@@ -89,37 +91,79 @@ const SidebarItemList = styled('ul')`
...
     </FeatureGuard>
   );
 }
 
-const SidebarLink = styled(Link)`
+const StyledLink = styled(Link)`
   position: relative;
 `;
 
+const StyledButton = styled('button')`
```
Comment (evanpurkhiser) [inline static/app/components/nav/sidebar.tsx:121]: Any chance we could give this guy a better name, maybe `NavButton`?

I'm trying to highly discourage naming things `StyledX`

File: static/app/components/nav/utils.tsx (hunk 8 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx:131]: I can use a constant, but this value is never actually used. `URL` expects a fully formed URL string, so we need to pass something in as a base. `http://example.com/` would also work here.

File: static/app/components/nav/utils.tsx [hunk 12]
```diff
@@ -1,31 +1,49 @@
...
 import type Feature from 'sentry/components/acl/feature';
+import type {MenuItemProps} from 'sentry/components/dropdownMenu';
 import {isItemActive} from 'sentry/components/sidebar/sidebarItem';
 import {SIDEBAR_NAVIGATION_SOURCE} from 'sentry/components/sidebar/utils';
 import normalizeUrl from 'sentry/utils/url/normalizeUrl';
 import type {useLocation} from 'sentry/utils/useLocation';
 
-/** NavItem is the base class for both SidebarItem and SubmenuItem  */
+/**
+ ** NavItem is the base class for both SidebarItem and SubmenuItem
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:11]:
```suggestion
 * NavItem is the base class for both SidebarItem and SubmenuItem
```

File: static/app/components/nav/utils.tsx [hunk 13]
```diff
@@ -1,31 +1,49 @@
...
 interface NavItem {
-  /** User-facing item label, surfaced in the UI. Should be translated! */
+  /**
+   ** User-facing item label, surfaced in the UI. Should be translated!
+   */
   label: string;
-  /** Optionally, props which should be passed to a wrapping `<Feature>` guard */
-  feature?: Omit<ComponentProps<typeof Feature>, 'children'>;
+  /**
+   ** Optionally, props which should be passed to a wrapping `<Feature>` guard
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:19]:
```suggestion
   * Optionally, props which should be passed to a wrapping `<Feature>` guard
```

Comment (evanpurkhiser) [reply static/app/components/nav/utils.tsx:19]: Actually won't do it to all of them but you get the idea lol

File: static/app/components/nav/utils.tsx (hunk 8 above)
Comment (evanpurkhiser) [reply static/app/components/nav/utils.tsx:131]: hmmm, does this work ok for on premise and things like that?

File: static/app/components/nav/sidebar.tsx [hunk 14]
```diff
@@ -89,37 +91,79 @@ const SidebarItemList = styled('ul')`
...
+function SidebarMenu({item}: PropsWithChildren<SidebarItemProps>) {
+  return (
+    <DropdownMenu
+      position="right-end"
+      trigger={(props, isOpen) => {
+        return (
+          <StyledButton {...props}>
+            <InteractionStateLayer hasSelectedBackground={isOpen} />
+            {item.icon}
+            <span>{item.label}</span>
```
Comment (evanpurkhiser) [inline static/app/components/nav/sidebar.tsx:158]: Shouldn't this just render `{children}` since we pass these two things in the component above

File: static/app/components/nav/sidebar.tsx [hunk 15]
```diff
@@ -89,37 +91,79 @@ const SidebarItemList = styled('ul')`
...
-  const isSubmenuActive = isSubmenuItemActive(item, location);
-  const _to = resolveNavItemTo(item);
-  const linkProps = _to ? makeLinkPropsFromTo(_to) : {to: '#'};
+interface SidebarItemProps {
+  item: NavSidebarItem;
+}
+
+function SidebarItem({item}: SidebarItemProps) {
+  const to = resolveNavItemTo(item);
+  const Component = to ? SidebarLink : SidebarMenu;
```
Comment (evanpurkhiser) [inline static/app/components/nav/sidebar.tsx:100]: @JonasBa do you have any strong feelings on this pattern? It makes it a little trickier for typescript, vs if we did something like

```suggestion
  const item = to ? (
    <SidebarLink item={item} key={item.label} />
  ) : (
    <SidebarMenu item={item} key={item.label} />
  );
```

then typescript works a little less hard. In this case though they would need to be responsible for rendering their children

File: static/app/components/nav/utils.tsx [hunk 16]
```diff
@@ -1,31 +1,49 @@
...
+   */
+  icon: React.ReactElement;
+  /**
+   ** if te items to display when this SidebarItem is active
+   */
+  menu?: MenuItemProps[];
+  /**
+   ** Optionally, the submenu items to display when this SidebarItem is active
+   */
   submenu?: NavSubmenuItem[] | NavItemLayout<NavSubmenuItem>;
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:47]: I wonder if these two are confusing now? Maybe a different name for one of these might make sense

File: static/app/components/nav/overlay.tsx
Comment (natemoo-re) [reply static/app/components/nav/overlay.tsx:1]: Good catch, removed in f1efc6c6948

File: static/app/components/nav/overlay.tsx (hunk 7 above)
Comment (natemoo-re) [reply static/app/components/nav/overlay.tsx]: This file was from an earlier version of the PR! Removed in f1efc6c6948

File: static/app/components/nav/overlay.tsx (hunk 6 above)
Comment (natemoo-re) [reply static/app/components/nav/overlay.tsx]: Point taken! File (and all new components) removed in f1efc6c6948

File: static/app/components/nav/utils.tsx (hunk 16 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx:47]: Oh good call, totally missed that they are too similar now

File: static/app/components/nav/sidebar.tsx (hunk 15 above)
Comment (JonasBa) [reply static/app/components/nav/sidebar.tsx:100]: It's not an uncommon pattern tbh. The drawback of this alternative is that the render body ends up bit less readable imo, as {item} doesnt convey that this is a component and could be any value. I tend to prefer to use this approach here, but if the type of component is limited I might just call it something else than `Component` to be more descriptive

```jsx
return (
	<div>
		{item}
		<Component ... />
   </div>
)
```

File: static/app/components/nav/overlay.tsx (hunk 4 above)
Comment (natemoo-re) [reply static/app/components/nav/overlay.tsx]: Good call! I will definitely keep that in mind. I refactored another usage in this area in 7c73aeaf47d

File: static/app/components/nav/sidebar.tsx (hunk 9 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: That's annoying, thanks for the callout!

File: static/app/components/nav/sidebar.tsx (hunk 11 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: Updated in e628dcdb963

File: static/app/components/nav/utils.tsx (hunk 16 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx]: Fixed in 656b978071c

File: static/app/components/nav/sidebar.tsx [hunk 17]
```diff
@@ -89,37 +91,79 @@ const SidebarItemList = styled('ul')`
...
+      trigger={(props, isOpen) => {
+        return (
+          <StyledButton {...props}>
+            <InteractionStateLayer hasSelectedBackground={isOpen} />
+            {item.icon}
+            <span>{item.label}</span>
+          </StyledButton>
+        );
+      }}
+      items={item.menu!}
```
[the remark this answers is not in the document]
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: Makes sense—fixed this in 656b978071c

File: static/app/components/nav/sidebar.tsx (hunk 14 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: Good catch, refactored in 9580e0c83f2!

File: static/app/components/nav/sidebar.tsx (hunk 15 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx]: Thanks for kicking off this conversation @evanpurkhiser! I agree with @JonasBa that the current `Component` pattern is well-suited here, since both component types expect identical props, children, and only differ in implementation. I did refactor to a more specific name in 9580e0c83f2. 

Would love to discuss possible TypeScript performance optimizations we can make across the codebase at the next FE TSC!

File: static/app/components/nav/utils.tsx (hunk 8 above)
Comment (natemoo-re) [reply static/app/components/nav/utils.tsx:134]: I added a comment above this code to explain more! 

Also made sure to pass this through `normalizeUrl()` to address on-prem concerns. See 64ab49c1d2c.

File: static/app/components/nav/sidebar.tsx [hunk 18]
```diff
@@ -89,37 +91,83 @@ const SidebarItemList = styled('ul')`
...
+    </NavLink>
+  );
+}
+
+function SidebarMenu({item, children}: SidebarItemProps & {children: React.ReactNode}) {
+  if (!item.dropdown) {
+    throw new Error(
+      `Nav item "${item.label}" must have either a \`dropdown\` or \`to\` value!`
+    );
+  }
```
Comment (evanpurkhiser) [inline static/app/components/nav/sidebar.tsx:154]: Should there be a similar check in SiedbarLink?

File: static/app/components/nav/utils.tsx [hunk 19]
```diff
@@ -1,31 +1,49 @@
...
+import type {MenuItemProps} from 'sentry/components/dropdownMenu';
 import {isItemActive} from 'sentry/components/sidebar/sidebarItem';
 import {SIDEBAR_NAVIGATION_SOURCE} from 'sentry/components/sidebar/utils';
 import normalizeUrl from 'sentry/utils/url/normalizeUrl';
 import type {useLocation} from 'sentry/utils/useLocation';
 
-/** NavItem is the base class for both SidebarItem and SubmenuItem  */
+/**
+ * NavItem is the base class for both SidebarItem and SubmenuItem
+ */
```
Comment (evanpurkhiser) [inline static/app/components/nav/utils.tsx:12]: nit: would be nice to see these in a different PR since (these are changes!) this clutters up this PR a bit

**Review by evanpurkhiser — APPROVED** [review]

File: static/app/components/nav/sidebar.tsx (hunk 18 above)
Comment (natemoo-re) [reply static/app/components/nav/sidebar.tsx:159]: Good call, added in [b5e7e44](https://github.com/getsentry/sentry/pull/79720/commits/b5e7e44da2e819b015e0e99f20db5aa8f52e1473)
